erpnext/controllers/selling_controller.py

```diff
@@ -55,11 +55,25 @@ def validate_coupon(self):
 		if self.meta.get_field("coupon_code"):
 
 			if self.coupon_code:
-				coupon_customer = frappe.get_value("Coupon Code", self.coupon_code, "allowed_customer")
-				coupon_customer_group = frappe.get_value("Coupon Code", self.coupon_code, "allowed_customer_group")
+				coupon = frappe.get_doc("Coupon Code", self.coupon_code)
+
+				coupon_customer = coupon.allowed_customer
+				coupon_customer_group = coupon.allowed_customer_group
+
+				if coupon.valid_from:
+					if coupon.valid_from > getdate(today()):
+						frappe.throw(_("Sorry, this coupon code's validity has not started"), title=_("Coupon Error"))
+				elif coupon.valid_upto:
+					if coupon.valid_upto < getdate(today()):
+						frappe.throw(_("Sorry, this coupon code's validity has expired"), title=_("Coupon Error"))
+				elif coupon.used >= coupon.maximum_use:
+					frappe.throw(_("Sorry, this coupon code is no longer valid"), title=_("Coupon Error"))
 
 				party_name = self.customer_name if self.doctype != "Quotation" else self.party_name
 
+				if not coupon.enabled:
+					frappe.throw(_("Coupon \"{}\" is not a valid coupon").format(self.coupon_code))
```

Comment:
I think we should make this as the first check, it should come before date validations